File: testdata/fixtures/csharp/CWE-352.cs
// CWE-352: CSRF — POST actions without ValidateAntiForgeryToken.
// ASP.NET validates CSRF tokens with [ValidateAntiForgeryToken]; POST
// handlers without it can be forged from another origin.
using Microsoft.AspNetCore.Mvc;

namespace Fixture {
    public class TransferController : Controller {
        // ruleid: csharp-csrf
        [HttpPost]
        public IActionResult Transfer() {
            return Ok();
        }

        // ruleid: csharp-csrf
        [HttpPost]
        public IActionResult Update() {
            return Ok();
        }

        // ruleid: csharp-csrf
        [HttpPost]
        public IActionResult Delete() {
            return Ok();
        }

        // ok: csharp-csrf
        [HttpPost, ValidateAntiForgeryToken] public IActionResult SafeTransfer() {
            return Ok();
        }

        // ok: csharp-csrf
        [HttpPost, ValidateAntiForgeryToken] public IActionResult SafeUpdate() {
            return Ok();
        }
    }
}

File: testdata/fixtures/go/CWE-352.go
// CWE-352: CSRF — Go handlers without protection.
// Handlers registered with http.HandleFunc that are not wrapped by a CSRF
// middleware can be forged from another origin. Safe handlers use a
// protection wrapper.
package fixture

import (
	"net/http"
)

func vuln1() {
	// ruleid: go-csrf-missing
	http.HandleFunc("/transfer", func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, "/done", http.StatusFound)
	})
}

func vuln2() {
	// ruleid: go-csrf-missing
	http.HandleFunc("/update", func(w http.ResponseWriter, r *http.Request) {
		r.ParseForm()
	})
}

func vuln3() {
	// ruleid: go-csrf-missing
	http.HandleFunc("/sensitive", func(w http.ResponseWriter, r *http.Request) {
	})
}

// ok helpers — not http handlers, so not flagged
func helperNotHandler() {
	// ok: go-csrf-missing
	println("not a handler")
}

// ok: go-csrf-missing
func anotherHelper() {
	// not a HandleFunc
}
